### PR TITLE
Orbital optimization refactor

### LINKED ISSUE #70: Add orbital-optimized DOCI reference energies
Cristina has a reference energy for H2@OO-DOCI//6-31G**: `-1.1651486697`. The geometry is said to be
> 2
> 
> H       0.0     0.0     0.371000028
> H       0.0     0.0     -0.371000028

### LINKED ISSUE #138: Clean up OO-DOCI
Some changes would be beneficial for the OO-DOCI code:

- [ ] calculate the electronic gradient in a separate function
- [ ] calculate the electronic Hessian in a separate function

Consider, too, if it is possible to generalize the algorithms in OO-DOCI and ERNewtonLocalizer.

Furthermore, shouldn't
```cpp
Eigen::VectorXd kappa_vector = numopt::newtonStep(Eigen::VectorXd::Zero(K), gradient_function, hessian_function);  // with only the free parameters
```
be 
```cpp
Eigen::VectorXd::Zero(K*(K-1)/2)
```
? However, it doesn't really matter, since `gradient_function` and `hessian_function` return a gradient and Hessian with the correct dimension.

### LINKED ISSUE #143: Orbital optimization - generalizations
There are big similarities between the orbital optimizers (OO-DOCI, OO-AP1roG) and the orbital localizers (Edmiston-Ruedenberg, Boys, ...): they all **optimize** (minimize/maximize) **a cost function** (DOCI energy, AP1roG energy, ER localization index, Boys localization index) by information of either:
- the value of the cost function, as a function of the Jacobi rotation parameters
- the behavior of the cost function near the current point: Newton-based(gradient and Hessian) or even gradient-ascent-based

There should be a way to generalize these algorithms, possibly even in numopt. Some initial thoughts:
- Could there be a `CostFunction` class, or `CostFunctionParameter`, which abstracts the cost function and its parameters?
-  How does simulated annealing fit into this frame of thought?

### PR DESCRIPTION
**Pull request by lelemmen**
In this PR, I will refactor the orbital optimization algorithms in order to remove duplicate code. More specifically, it:
- [x] closes #70 
- [x] closes #138 
- [x] closes #143

### CODE REVIEW

**Comment by lelemmen** [comment]
@guacke @tmhuysen It would be nice if both of you gave your comments on this OO refactor PR.

File: include/OrbitalOptimization/OrbitalOptimizationOptions.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+// 
+// GQCG-gqcp is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-gqcp.  If not, see <http://www.gnu.org/licenses/>.
+// 
+#ifndef OrbitalOptimizationOptions_hpp
```
Comment (tmhuysen) [inline include/OrbitalOptimization/OrbitalOptimizationOptions.hpp]: idem

File: include/OrbitalOptimization/OrbitalOptimizationOptions.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#define OrbitalOptimizationOptions_hpp
```
Comment (tmhuysen) [inline include/OrbitalOptimization/OrbitalOptimizationOptions.hpp]: idem

File: src/OrbitalOptimization/OrbitalOptimizationOptions.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ *  @param maximum_number_of_iterations     the maximum number of iterations that may be used to achieve convergence
+ * 
+ *  @return orbital optimization options suited for the minimization of a cost function
+ */
+OrbitalOptimizationOptions OrbitalOptimizationOptions::OrbitalMinimizationOptions(const double convergence_threshold, const double maximum_number_of_iterations) {
+    return OrbitalOptimizationOptions(true, convergence_threshold, maximum_number_of_iterations);
+}
+
+
+}  // namespace GQCP
```
Comment (tmhuysen) [inline src/OrbitalOptimization/OrbitalOptimizationOptions.cpp]: add line

File: include/OrbitalOptimization/JacobiOrbitalOptimizer.hpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+#include "OrbitalOptimization/BaseOrbitalOptimizer.hpp"
+#include "OrbitalOptimization/JacobiRotationParameters.hpp"
+
+#include <utility>
+
+
+namespace GQCP {
+
+
+class JacobiOrbitalOptimizer : public BaseOrbitalOptimizer {
```
Comment (tmhuysen) [inline include/OrbitalOptimization/JacobiOrbitalOptimizer.hpp]: There is no class discription

**Comment by lelemmen** [comment]
@tmhuysen Thanks for the review!

**Pull request merged**